pymilvus/milvus_client/milvus_client.py

```diff
@@ -0,0 +1,801 @@
+        self.collection_name = collection_name
+        self.shard_num = shard_num
+        self.partitions = partitions
+        self.consistency_level = consistency_level
+        self.replica_number = replica_number
+        self.index_params = index_params
+        self.timeout = timeout
+        self.pk_field = pk_field
+        self.vector_field = vector_field
+
+        # TODO: Figure out thread safety
+        # self.concurrent_counter = 0
+        self.concurrent_lock = threading.RLock()
+        self.dim = None
+        self.default_search_params = None
+        self.collection = None
+        self.fields = None
+
+        self.alias = self._create_connection()
+        self.is_self_hosted = bool(
+            utility.get_server_type(using=self.alias) == "milvus"
+        )
+        if drop_old:
+            self.delete_collection()
+        self._init(None)
+
+    def insert_data(
+        self,
+        data: List[Dict[str, any]],
+        timeout: int = None,
+        batch_size: int = 100,
+        partition: str = None,
+        progress_bar: bool = False,
+    ) -> List[Union[str, int]]:
+        """Insert data into the collection.
+
+        If the Milvus Client was initiated without an existing Collection, the first dict passed
+        in will be used to initiate the collection.
+
+        Args:
+            data (List[Dict[str, any]]): A list of dicts to pass in. If list not provided, will
+                cast to list.
+            timeout (int, optional): The timeout to use, will override init timeout. Defaults
+                to None.
+            batch_size (int, optional): The batch size to perform inputs with. Defaults to 100.
+            partition (str, optional): Which partition to insert into. Defaults to None.
+            progress_bar (bool, optional): Whether to display a progress bar for the input.
+                Defaults to False.
+
+        Raises:
+            DataNotMatchException: If the data has misssing fields an exception will be thrown.
+            MilvusException: General Milvus error on insert.
+
+        Returns:
+            List[Union[str, int]]: A list of primary keys that were inserted.
+        """
+        # If no data provided, we cannot input anything
+        if len(data) == 0:
+            return []
+
+        # If the collection hasnt been initialized, initialize it
+        with self.concurrent_lock:
+            if self.collection is None:
+                self._init(data[0])
+
+        # Dont include the primary key if auto_id is true and they included it in data
+        ignore_pk = self.pk_field if self.collection.schema.auto_id else None
+        insert_dict = {}
+        pks = []
+
+
+        for k in data:
+            for key, value in k.items():
+                if key in self.fields:
+                    insert_dict.setdefault(key, []).append(value)
+
+        # Insert the data in batches
+        for i in tqdm(range(0, len(data), batch_size), disable=not progress_bar):
```

Comment:
I love the batch design but we need to test the corner case.
For ex. batch_size = 0

Reply:
good catch